lib/parser.mly

```diff
@@ -43,14 +44,18 @@ let stmt :=
   | LPAREN; GET_MODEL; RPAREN; { Ast.GetModel }
 
 let s_expr :=
-  | x = SYMBOL; { mk_symbol_s (get_bind x) x }
-  | c = spec_constant; { Val c }
-  | LPAREN; op = UNARY; e = s_expr; RPAREN; { Unop (op, e) }
-  | LPAREN; op = BINARY; e1 = s_expr; e2 = s_expr; RPAREN; { Binop (op, e1, e2) }
-  | LPAREN; op = TERNARY; e1 = s_expr; e2 = s_expr; e3 = s_expr; RPAREN;
-    { Triop (op, e1, e2, e3) }
-  | LPAREN; op = CVTOP; e = s_expr; RPAREN; { Cvtop (op, e) }
-  | LPAREN; op = RELOP; e1 = s_expr; e2 = s_expr; RPAREN; { Relop (op, e1, e2) }
+  | x = SYMBOL; { mk_symbol @@ Symbol.mk_symbol (get_bind x) x }
+  | c = spec_constant; { { ty = Value.type_of c; e = Val c } }
+  | LPAREN; (ty, op) = UNARY; e = s_expr; RPAREN;
+    { { ty; e = Unop (op, e) } }
+  | LPAREN; (ty, op) = BINARY; e1 = s_expr; e2 = s_expr; RPAREN;
+    { { ty; e = Binop (op, e1, e2) } }
+  | LPAREN; (ty, op) = TERNARY; e1 = s_expr; e2 = s_expr; e3 = s_expr; RPAREN;
+    { { ty; e = Triop (op, e1, e2, e3) } }
+  | LPAREN; (ty, op) = CVTOP; e = s_expr; RPAREN;
+    { { ty; e = Cvtop (op, e) } }
+  | LPAREN; (ty, op) = RELOP; e1 = s_expr; e2 = s_expr; RPAREN;
```

Comment:
It should be possible to move all the `LPAREN; ...; RPAREN;` to their own rule `paren_op` for instance and have a single `LPAREN; op = paren_op; RPAREN`.

Reply:
Oh yes, nice thinking! It makes the parsing rules more concise :grin: